Stack/Opc.Ua.Core/Security/Certificates/Nonce.cs

```diff
@@ -0,0 +1,470 @@
+
+                byte[] output = new byte[length];
+
+                HMAC hmac = returnHMACInstance(secret, algorithm);
+
+                byte counter = 1;
+
+                byte[] info = new byte[hmac.HashSize / 8 + salt.Length + 1];
+                Buffer.BlockCopy(salt, 0, info, 0, salt.Length);
+                info[salt.Length] = counter++;
+
+                byte[] hash = hmac.ComputeHash(info, 0, salt.Length + 1);
+
+                int pos = 0;
+
+                for (int ii = 0; ii < hash.Length && pos < length; ii++)
+                {
+                    output[pos++] = hash[ii];
+                }
+
+                while (pos < length)
+                {
+                    Buffer.BlockCopy(hash, 0, info, 0, hash.Length);
+                    Buffer.BlockCopy(salt, 0, info, hash.Length, salt.Length);
+                    info[info.Length - 1] = counter++;
+
+                    hash = hmac.ComputeHash(info, 0, info.Length);
+
+                    for (int ii = 0; ii < hash.Length && pos < length; ii++)
+                    {
+                        output[pos++] = hash[ii];
+                    }
+                }
+
+                return output;
+            }
+
+            return Data;
+        }
+
+        /// <summary>
+        /// Creates a nonce for the specified security policy URI and nonce length.
+        /// </summary>
+        /// <param name="securityPolicyUri">The security policy URI.</param>
+        /// <param name="nonceLength">The length of the nonce.</param>
+        /// <returns>A <see cref="Nonce"/> object containing the generated nonce.</returns>
+        public static Nonce CreateNonce(string securityPolicyUri, uint nonceLength)
+        {
+            if (securityPolicyUri == null)
+            {
+                throw new ArgumentNullException(nameof(securityPolicyUri));
+            }
+
+            Nonce nonce = null;
+
+            switch (securityPolicyUri)
+            {
+                case SecurityPolicies.ECC_nistP256: { return CreateNonce(ECCurve.NamedCurves.nistP256); }
+                case SecurityPolicies.ECC_nistP384: { return CreateNonce(ECCurve.NamedCurves.nistP384); }
+                case SecurityPolicies.ECC_brainpoolP256r1: { return CreateNonce(ECCurve.NamedCurves.brainpoolP256r1); }
+                case SecurityPolicies.ECC_brainpoolP384r1: { return CreateNonce(ECCurve.NamedCurves.brainpoolP384r1); }
+#if CURVE25519
+                case SecurityPolicies.ECC_curve25519:
+                {
+                    return CreateNonceForCurve25519();
+                }
+
+                case SecurityPolicies.ECC_curve448:
+                {
+                    return CreateNonceForCurve448();
+                }
+#endif
+                default:
+                {
+                    nonce = new Nonce() {
+                        Data = Utils.Nonce.CreateNonce(nonceLength)
```

Comment:
this is an opportunity to unify the Nonce class with Utils.Nonce and get rid of the legacy implementation. There should also be some tests for the new class.